File: 03-spring-examples/helpers/mock-external-service/src/main/java/com/nalepa/demo/MockExternalServiceEndpoint.java
package com.nalepa.demo;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;

import java.time.Duration;
import java.time.LocalTime;

@RestController
public class MockExternalServiceEndpoint {

    public record SomeResponse(String text) {}

    @GetMapping("/mock/{index}/{delaySeconds}")
    public ResponseEntity<SomeResponse> endpoint(
            @PathVariable String index,
            @PathVariable long delaySeconds
    ) {
        ResponseEntity<SomeResponse> response = ResponseEntity.ok(new SomeResponse("text"));
        nonBlockingSleep(delaySeconds);
        log(this, "Index: " + index + ". Response returned after seconds: " + delaySeconds);
        return response;
    }

    private static void log(Object caller, String message) {
        System.out.println(LocalTime.now() + " : " + caller.getClass().getSimpleName() + " : " + Thread.currentThread().getName() + " ### " + message);
    }

    private static void nonBlockingSleep(long seconds) {
        if (Thread.currentThread().isVirtual()) {
            try {
                Thread.sleep(Duration.ofSeconds(seconds));
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        } else {
            throw new RuntimeException("nonBlockingSleep called from carrier thread");
        }
    }
}

File: 03-spring-examples/helpers/mock-external-service/src/main/kotlin/com/nalepa/demo/MockExternalServiceEndpoint.kt
package com.nalepa.demo

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController
import java.time.Duration
import java.time.LocalTime

data class SomeResponse(
    val text: String,
)


@RestController
class MockExternalServiceEndpoint {

    @GetMapping("/mock/{index}/{delaySeconds}")
    fun endpoint(
        @PathVariable index: String,
        @PathVariable delaySeconds: Long,
    ): ResponseEntity<SomeResponse> =
        ResponseEntity
            .ok(SomeResponse("text"))
            .also {
                nonBlockingSleep(delaySeconds)
                MockExternalServiceLogger.log(this, "Index: $index. Response returned after seconds: $delaySeconds")
            }
}

object MockExternalServiceLogger {
    fun log(caller: Any, message: String) {
        println("${LocalTime.now()} : ${caller.javaClass.simpleName} : ${Thread.currentThread().name} ### $message")
    }
}

fun nonBlockingSleep(seconds: Long) {
    if (Thread.currentThread().isVirtual) {
        Thread.sleep(Duration.ofSeconds(seconds))
    } else {
        throw RuntimeException("nonBlockingSleep called from carrier thread")
    }
}